### PR TITLE
[Doc] Add Structured Generation tutorial and Quick Start

### PR DESCRIPTION
**Pull request by CharlieFRuan**
This PR adds `Structured Generation` tutorial with the following sections:
- JSON Generation
- JSON Schema Guided Generation
- EBNF Grammar Guided Generation
- Structured Generation for Batched Inference

This PR also adds `Quick Start` that has a simplified code snippet of JSON generation.

This PR also fixes a minor bug in the revamped API, where we did not pass `index` in `GrammarMatcher_FillNextTokenBitmask()` for `python_methods.cc`.

### CODE REVIEW

**Comment by tqchen** [comment]
Update to latest api and let us merge iterate

File: docs/README.md [hunk 1]
```diff
@@ -27,4 +27,4 @@ cd _build/html
 python3 -m http.server
 ```
 
-Then you can view the documentation in your browser at `http://localhost:8000` (the port can be customized by appending ` -p PORT_NUMBER` in the python command above).
+Then you can view the documentation in your browser at `http://localhost:8000` (the port can be customized by appending ` -p PORT_NUMBER` in the python command above, `--bind 0.0.0.0` if needed.).
```
Comment (Ubospica) [inline docs/README.md]: What is the purpose of this binding?

**Review by Ubospica — COMMENTED** [review]
Thanks for the contribution! Looks great.

File: docs/README.md (hunk 1 above)
Comment (CharlieFRuan) [reply docs/README.md]: From my experience on Mac, without this I could not open the doc with `localhost`

Comment (Ubospica) [reply docs/README.md]: Could you add this in doc?

Comment (CharlieFRuan) [reply docs/README.md]: You mean in this README.md right

Comment (Ubospica) [reply docs/README.md]: Yes

Comment (CharlieFRuan) [reply docs/README.md]: Added

Comment (Ubospica) [reply docs/README.md]: Then feel free to merge it

**Pull request merged**